### ISSUE TITLE
niftyreg interfaces failing because of base and OMP_NUM_THREADS

### ISSUE THREAD

**Issue by byvernault** [nipy/nipype#2004]
### Summary

I tried to install the master version of nipype and run our workflows using niftyreg and niftyseg. It looks like the node using niftyreg/niftyseg are not working because of the code to set the OMP_NUM_THREADS in the _run_interface.

Right now, the code is the following:

if not isdefined(self.inputs.environ['OMP_NUM_THREADS']):
            self.inputs.environ['OMP_NUM_THREADS'] = self.num_threads

But the self.inputs.environ is not populated from the class CommandLine giving us the error  below.
 
I could fix it by changing to this:

if 'OMP_NUM_THREADS' not in os.environ:
            os.environ['OMP_NUM_THREADS'] = str(self.num_threads)

or do you recommend something else? Where should self.inputs.environ be set?

### Actual behavior

Error when running a niftyreg node:

170510-11:56:37,203 workflow INFO:
[2017-05-10 11:56:37] 	 Traceback (most recent call last):
[2017-05-10 11:56:37]   File "/Users/byvernault/home-local/code/anaconda/lib/python2.7/site-packages/nipype/pipeline/plugins/linear.py", line 43, in run
[2017-05-10 11:56:37]     node.run(updatehash=updatehash)
[2017-05-10 11:56:37]   File "/Users/byvernault/home-local/code/anaconda/lib/python2.7/site-packages/nipype/pipeline/engine/nodes.py", line 372, in run
[2017-05-10 11:56:37]     self._run_interface()
[2017-05-10 11:56:37]   File "/Users/byvernault/home-local/code/anaconda/lib/python2.7/site-packages/nipype/pipeline/engine/nodes.py", line 1287, in _run_interface
[2017-05-10 11:56:37]     updatehash=updatehash))
[2017-05-10 11:56:37]   File "/Users/byvernault/home-local/code/anaconda/lib/python2.7/site-packages/nipype/pipeline/engine/nodes.py", line 1193, in _collate_results
[2017-05-10 11:56:37]     (self.name, '\n'.join(msg)))
[2017-05-10 11:56:37] Exception: Subnodes of node: register2GAD failed:
[2017-05-10 11:56:37] Subnode 0 failed
[2017-05-10 11:56:37] Error:
[2017-05-10 11:56:37] u'OMP_NUM_THREADS\nInterface RegAladin failed to run. '

### How to replicate the behavior

Run an interface using niftyreg like follow:

import nipype.interfaces.niftyreg as niftyreg

test_node = niftyreg.RegAladin()
test_node.base_dir = '~/data/test_reg'
test_node.inputs.flo_file = 'image1.nii.gz'
test_node.inputs.ref_file = 'image1.nii.gz'

print test_node.cmdline

result = test_node.run()

print result.outputs

### Platform details:

{'nibabel_version': '2.1.0', 'sys_executable': '/Users/byvernault/home-local/code/anaconda/bin/python', 'networkx_version': '1.11', 'numpy_version': '1.12.1', 'sys_platform': 'darwin', 'sys_version': '2.7.12 |Anaconda 4.2.0 (x86_64)| (default, Jul  2 2016, 17:43:17) \n[GCC 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)]', 'commit_source': u'installation', 'commit_hash': u'bfdb164f2', 'pkg_path': '/Users/byvernault/home-local/code/anaconda/lib/python2.7/site-packages/nipype-0.13.0_gbfdb164f2.dev-py2.7.egg/nipype', 'nipype_version': u'0.13.0-dev', 'traits_version': '4.6.0', 'scipy_version': '0.19.0'}
0.13.0-dev

### Execution environment

- My python environment outside container

**Comment by byvernault** [comment]
I added:

self.inputs.environ = dict(os.environ)

before the if statement and it's working.

**Comment by effigies** [comment]
```Python
if not isdefined(self.inputs.environ['OMP_NUM_THREADS']):
    self.inputs.environ['OMP_NUM_THREADS'] = self.num_threads
```

I think your problem is you're testing whether a dictionary entry is defined. What you probably want is:

```Python
if not isdefined(self.inputs.environ):
    self.inputs.environ = {}
self.inputs.environ.set_default('OMP_NUM_THREADS', self.num_threads)
```

This creates the dictionary if missing, and sets the value if missing. If it exists and `OMP_NUM_THREADS` is set, it will have no effect, which I believe is what you want.

**Comment by byvernault** [comment]
Kind of. The issue here is that my self.inputs.environ is not populated from my env variables.

Shouldn't it be populate before _run_interface()?

Right now, your changes will always set OMP_NUM_THREADS to self.num_threads since self.inputs.environ doesn't read my env variables.

An other fix is to go directly to os.environ to do this:

if 'OMP_NUM_THREADS' not in os.environ:
    os.environ['OMP_NUM_THREADS'] = str(self.num_threads)

WDYT?

**Comment by effigies** [comment]
> Kind of. The issue here is that my self.inputs.environ is not populated from my env variables.

This is a decision of the workflow writer. If I choose to set:

```Python
x = pe.Node(Interface(), name='x')
x.inputs.environ = os.environ
```

Then fine. But I don't think an interface should make that assumption. (Someone feel free to step in if you can think of an exception.) Consider if I carefully set the environment, and then the interface overrides it with the host environment: I no longer am controlling the context of my computation.

> Shouldn't it be populate before _run_interface()?

By doing it then, people can update environment or `num_threads` entries at any point prior to running and be sure that it will be applied correctly.

> Right now, your changes will always set OMP_NUM_THREADS to self.num_threads since self.inputs.environ doesn't read my env variables.

You can set:

```Python
x.inputs.num_threads = int(os.environ['OMP_NUM_THREADS'])
```

> An other fix is to go directly to os.environ to do this:
> 
> if 'OMP_NUM_THREADS' not in os.environ:
> os.environ['OMP_NUM_THREADS'] = str(self.num_threads)
> 
> WDYT?

This would affect every single interface that tries to read `os.environ['OMP_NUM_THREADS']` after this line is run. I'd try to avoid modifying global state like that.

**Comment by byvernault** [comment]
I see. Ok, I will edit it then your way and do a push request. 

Thank you very much @effigies

**Comment by effigies** [comment]
Closed by #2005.
